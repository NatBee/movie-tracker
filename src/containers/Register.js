import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setUser} from '../actions/index';
import { userSignIn, addNewUser } from '../apiCall'



class Register extends Component {
  constructor(props) {
    super(props);
    
    this.state ={
      name: '',
      email: '',
      password: '',
    }
  }

  handleChange = (e) => {
    e.preventDefault();
    const field = e.target.className;
    const value = e.target.value;

    this.setState({ [field]: value });
  }

  registerSubmit = async (e) => {

    const { name, email, password } = this.state

    e.preventDefault();
    const results = await addNewUser(name, email, password);
    const userResults = await userSignIn(email, password);
    const user = await userResults.data
    this.props.setUser(user)
  }

  render() {
    return (
      <div className="log-in-form">
        <form onSubmit={this.registerSubmit}>
          <input type="text" className="name" placeholder="Your Name" name="email" onChange={this.handleChange}/>
          <input type="text" className="email" placeholder="email address" name="email" onChange={this.handleChange}/>
          <input type="password" className="password" placeholder="password" name="password" onChange={this.handleChange}/>
          <input type="submit" />
        </form>
      </div>
    )
  }
}

// export const mapStateToProps = (store) => ({
//   name: store.user.name
// })

export const mapDispatchToProps = (dispatch) => ({
  addNewUser: (user) => dispatch(addNewUser(user)),
})

export default connect(null, mapDispatchToProps)(Register)


