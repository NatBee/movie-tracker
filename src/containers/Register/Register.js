import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter, Link } from 'react-router-dom';
import { setUser } from '../../actions/index';
import { userSignIn, addNewUser } from '../../helper/apiCall';
import PropTypes from 'prop-types';

export class Register extends Component {
  constructor(props) {
    super(props);
    
    this.state ={
      name: '',
      email: '',
      password: '',
      errorMessage: null,
    }
  }

  handleChange = (e) => {
    e.preventDefault();
    const field = e.target.className;
    const value = e.target.value;

    this.setState({ [field]: value });
  }

  registerSubmit = async (e) => {
    e.preventDefault();
    
    const { name, email, password } = this.state
    const results = await addNewUser({ name, email, password });
    this.updateMessage(results)
  }

  updateMessage = async (results) => {
    if(results.status !== 'success') {
      this.setState({errorMessage: "That email is already registered. Please try again or login."})
    } else {
      this.autoSignIn();
    }
  }

  autoSignIn = async () => {
    const userEmail = this.state.email;
    const userPassword = this.state.password;

    const userResults = await userSignIn(userEmail, userPassword);
    const user = await userResults.data; 
    this.props.setUser(user)
    this.props.history.push('/')
  }

  render() {
    return (
      <div className="log-in-form">
        <form onSubmit={ this.registerSubmit }>
          <h3 className="acct-create">{ this.state.newAcctMessage }</h3>
          <input type="text" className="name" placeholder="Your Name" name="email" onChange={this.handleChange}/>
          <input type="text" className="email" placeholder="email address" name="email" onChange={this.handleChange}/>
          <input type="password" className="password" placeholder="password" name="password" onChange={this.handleChange}/>
          <button type="submit"
                  className="submit-button">Register</button>
          <h3 className="login-error">{ this.state.errorMessage }</h3>
          <Link to="/login">Login</Link>
        </form>
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  addNewUser: (user) => dispatch(addNewUser(user)),
  setUser: (user) => dispatch(setUser(user))
})

Register.propTypes = {
  history: PropTypes.object,
  addNewUser: PropTypes.func,
  setUser: PropTypes.func,
  user: PropTypes.object,
  registerSubmit: PropTypes.func,
  updateMessage: PropTypes.func,
  autoSignIn: PropTypes.func,
}

export default withRouter(connect(null, mapDispatchToProps)(Register));