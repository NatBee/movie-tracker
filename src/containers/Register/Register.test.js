/* eslint-disable */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { Register, mapDispatchToProps } from './Register';
import { Link } from 'react-router-dom';

describe('Register', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Register />)
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call dispatch method when using a function in mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);

    mapped.addNewUser();
    mapped.setUser();

    expect(mockDispatch).toHaveBeenCalled();
  });

  it.skip('should set state if registration is successful', async () => {
    wrapper = shallow(<Register />)
    const mockEvent = { preventDefault: jest.fn() };
  
    window.fetch = jest.fn().mockImplementationOnce(() => Promise.resolve({
      json: () => Promise.resolve({
        status: 'success'
      })
    }))

    await wrapper.instance().registerSubmit(mockEvent);
    wrapper.update()

    expect(wrapper.state('newAcctMessage')).toEqual("Registration successful. Please login.");
  });

  it('should set state if registration is not successful', async () => {
    wrapper = shallow(<Register />)
    const mockEvent = { preventDefault: jest.fn() };
  
    window.fetch = jest.fn().mockImplementationOnce(() => Promise.resolve({
      json: () => Promise.resolve({
        status: 'failure'
      })
    }))

    await wrapper.instance().registerSubmit(mockEvent);
    wrapper.update()

    expect(wrapper.state('errorMessage')).toEqual("That email is already registered. Please try again or login.");
  });
})