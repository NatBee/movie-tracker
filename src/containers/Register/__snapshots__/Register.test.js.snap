// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Register />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
      >
        <h3
          className="acct-create"
        />
        <input
          className="name"
          name="email"
          onChange={[Function]}
          placeholder="Your Name"
          type="text"
        />
        <input
          className="email"
          name="email"
          onChange={[Function]}
          placeholder="email address"
          type="text"
        />
        <input
          className="password"
          name="password"
          onChange={[Function]}
          placeholder="password"
          type="password"
        />
        <button
          className="submit-button"
          type="submit"
        >
          Register
        </button>
        <h3
          className="login-error"
        />
        <Link
          replace={false}
          to="/login"
        >
          Login
        </Link>
      </form>,
      "className": "log-in-form",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="acct-create"
          />,
          <input
            className="name"
            name="email"
            onChange={[Function]}
            placeholder="Your Name"
            type="text"
          />,
          <input
            className="email"
            name="email"
            onChange={[Function]}
            placeholder="email address"
            type="text"
          />,
          <input
            className="password"
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
          />,
          <button
            className="submit-button"
            type="submit"
          >
            Register
          </button>,
          <h3
            className="login-error"
          />,
          <Link
            replace={false}
            to="/login"
          >
            Login
          </Link>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "acct-create",
          },
          "ref": null,
          "rendered": null,
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "name",
            "name": "email",
            "onChange": [Function],
            "placeholder": "Your Name",
            "type": "text",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "email",
            "name": "email",
            "onChange": [Function],
            "placeholder": "email address",
            "type": "text",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "password",
            "name": "password",
            "onChange": [Function],
            "placeholder": "password",
            "type": "password",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Register",
            "className": "submit-button",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Register",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
            "className": "login-error",
          },
          "ref": null,
          "rendered": null,
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Login",
            "replace": false,
            "to": "/login",
          },
          "ref": null,
          "rendered": "Login",
          "type": [Function],
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
        >
          <h3
            className="acct-create"
          />
          <input
            className="name"
            name="email"
            onChange={[Function]}
            placeholder="Your Name"
            type="text"
          />
          <input
            className="email"
            name="email"
            onChange={[Function]}
            placeholder="email address"
            type="text"
          />
          <input
            className="password"
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
          />
          <button
            className="submit-button"
            type="submit"
          >
            Register
          </button>
          <h3
            className="login-error"
          />
          <Link
            replace={false}
            to="/login"
          >
            Login
          </Link>
        </form>,
        "className": "log-in-form",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3
              className="acct-create"
            />,
            <input
              className="name"
              name="email"
              onChange={[Function]}
              placeholder="Your Name"
              type="text"
            />,
            <input
              className="email"
              name="email"
              onChange={[Function]}
              placeholder="email address"
              type="text"
            />,
            <input
              className="password"
              name="password"
              onChange={[Function]}
              placeholder="password"
              type="password"
            />,
            <button
              className="submit-button"
              type="submit"
            >
              Register
            </button>,
            <h3
              className="login-error"
            />,
            <Link
              replace={false}
              to="/login"
            >
              Login
            </Link>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "acct-create",
            },
            "ref": null,
            "rendered": null,
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "name",
              "name": "email",
              "onChange": [Function],
              "placeholder": "Your Name",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "email",
              "name": "email",
              "onChange": [Function],
              "placeholder": "email address",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "password",
              "name": "password",
              "onChange": [Function],
              "placeholder": "password",
              "type": "password",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Register",
              "className": "submit-button",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Register",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": null,
              "className": "login-error",
            },
            "ref": null,
            "rendered": null,
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Login",
              "replace": false,
              "to": "/login",
            },
            "ref": null,
            "rendered": "Login",
            "type": [Function],
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
